File: app/src/main/java/com/drdisagree/iconify/utils/overlay/compiler/DynamicCompiler.java
package com.drdisagree.iconify.utils.overlay.compiler;

import static com.drdisagree.iconify.utils.helper.Logger.writeLog;

import android.os.Build;
import android.util.Log;

import com.drdisagree.iconify.BuildConfig;
import com.drdisagree.iconify.common.Const;
import com.drdisagree.iconify.common.Resources;
import com.drdisagree.iconify.utils.FileUtil;
import com.drdisagree.iconify.utils.RootUtil;
import com.drdisagree.iconify.utils.SystemUtil;
import com.drdisagree.iconify.utils.helper.BinaryInstaller;
import com.drdisagree.iconify.utils.overlay.OverlayUtil;
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager;
import com.topjohnwu.superuser.Shell;

import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class DynamicCompiler {

    private static final String TAG = DynamicCompiler.class.getSimpleName();
    private static String mOverlayName = null;
    private static String mPackage = null;
    private static boolean mForce = false;
    private static final String[] mResource = new String[3];
    private static final JSONObject[] jsonResources = new JSONObject[3];

    public static boolean buildOverlay() throws IOException {
        return buildOverlay(true);
    }

    public static boolean buildOverlay(boolean force) throws IOException {
        mForce = force;

        try {
            JSONObject[] jsonObject = ResourceManager.getResources();

            Shell.cmd("mkdir -p " + Resources.BACKUP_DIR).exec();

            for (int i = 0; i < 3; i++) {
                jsonResources[i] = ResourceManager.generateJsonResource(jsonObject[i]);
            }

            Iterator<String> keys = jsonResources[0].keys();

            // Create overlay for each package
            while (keys.hasNext()) {
                mPackage = keys.next();
                for (int i = 0; i < 3; i++) {
                    mResource[i] = jsonResources[i].getString(mPackage)
                            .replace("'", "\"")
                            .replace("><", ">\n<");
                }
                mOverlayName = mPackage.equals(Const.FRAMEWORK_PACKAGE) ? "Dynamic1" : "Dynamic2";

                preExecute();
                moveOverlaysToCache();

                // Create AndroidManifest.xml
                if (createManifestResource(mOverlayName, Resources.TEMP_CACHE_DIR + "/" + mPackage + "/" + mOverlayName)) {
                    Log.e(TAG, "Failed to create Manifest for " + mOverlayName + "! Exiting...");
                    postExecute(true);
                    return true;
                }

                // Build APK using AAPT
                if (OverlayCompiler.runAapt(Resources.TEMP_CACHE_DIR + "/" + mPackage + "/" + mOverlayName)) {
                    Log.e(TAG, "Failed to build " + mOverlayName + "! Exiting...");
                    postExecute(true);
                    return true;
                }

                // ZipAlign the APK
                if (OverlayCompiler.zipAlign(Resources.UNSIGNED_UNALIGNED_DIR + "/" + mOverlayName + "-unsigned-unaligned.apk")) {
                    Log.e(TAG, "Failed to align " + mOverlayName + "-unsigned-unaligned.apk! Exiting...");
                    postExecute(true);
                    return true;
                }

                // Sign the APK
                if (OverlayCompiler.apkSigner(Resources.UNSIGNED_DIR + "/" + mOverlayName + "-unsigned.apk")) {
                    Log.e(TAG, "Failed to sign " + mOverlayName + "-unsigned.apk! Exiting...");
                    postExecute(true);
                    return true;
                }

                postExecute(false);
            }

            if (mForce) {
                Shell.cmd("rm -rf " + Resources.BACKUP_DIR).exec();

                // Disable the overlays incase they are already enabled
                OverlayUtil.disableOverlays("IconifyComponentDynamic1.overlay", "IconifyComponentDynamic2.overlay");

                // Install from files dir
                for (int i = 1; i <= 2; i++) {
                    Shell.cmd("pm install -r " + Resources.DATA_DIR + "/IconifyComponentDynamic" + i + ".apk").exec();
                    Shell.cmd("rm -rf " + Resources.DATA_DIR + "/IconifyComponentDynamic" + i + ".apk").exec();
                }

                // Move to system overlay dir
                SystemUtil.mountRW();
                for (int i = 1; i <= 2; i++) {
                    Shell.cmd("cp -rf " + Resources.SIGNED_DIR + "/IconifyComponentDynamic" + i + ".apk " + Resources.SYSTEM_OVERLAY_DIR + "/IconifyComponentDynamic" + i + ".apk").exec();
                    RootUtil.setPermissions(644, "/system/product/overlay/IconifyComponentDynamic" + i + ".apk");
                }
                SystemUtil.mountRO();

                // Enable the overlays
                OverlayUtil.enableOverlays("IconifyComponentDynamic1.overlay", "IconifyComponentDynamic2.overlay");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to build overlay! Exiting...", e);
            postExecute(true);
            return true;
        }

        return false;
    }

    private static void preExecute() throws IOException {
        // Create symbolic link
        BinaryInstaller.symLinkBinaries();

        // Clean data directory
        Shell.cmd("rm -rf " + Resources.TEMP_DIR).exec();
        Shell.cmd("rm -rf " + Resources.DATA_DIR + "/Overlays").exec();

        // Extract overlay from assets
        FileUtil.copyAssets("Overlays/" + mPackage + "/" + mOverlayName);

        // Create temp directory
        Shell.cmd("rm -rf " + Resources.TEMP_DIR + "; mkdir -p " + Resources.TEMP_DIR).exec();
        Shell.cmd("mkdir -p " + Resources.TEMP_OVERLAY_DIR).exec();
        Shell.cmd("mkdir -p " + Resources.TEMP_CACHE_DIR).exec();
        Shell.cmd("mkdir -p " + Resources.UNSIGNED_UNALIGNED_DIR).exec();
        Shell.cmd("mkdir -p " + Resources.UNSIGNED_DIR).exec();
        Shell.cmd("mkdir -p " + Resources.SIGNED_DIR).exec();
        Shell.cmd("mkdir -p " + Resources.TEMP_CACHE_DIR + "/" + mPackage + "/").exec();
        Shell.cmd("mkdir -p " + Resources.BACKUP_DIR).exec();
    }

    private static void postExecute(boolean hasErroredOut) {
        if (!hasErroredOut) {
            // Move all generated overlays to module
            Shell.cmd("cp -rf " + Resources.SIGNED_DIR + "/IconifyComponent" + mOverlayName + ".apk " + Resources.OVERLAY_DIR + "/IconifyComponent" + mOverlayName + ".apk").exec();
            RootUtil.setPermissions(644, Resources.OVERLAY_DIR + "/IconifyComponent" + mOverlayName + ".apk");

            Shell.cmd("cp -rf " + Resources.SIGNED_DIR + "/IconifyComponent" + mOverlayName + ".apk " + Resources.BACKUP_DIR + "/IconifyComponent" + mOverlayName + ".apk").exec();

            // Move to files dir
            if (mForce) {
                Shell.cmd("cp -rf " + Resources.SIGNED_DIR + "/IconifyComponent" + mOverlayName + ".apk " + Resources.DATA_DIR + "/IconifyComponent" + mOverlayName + ".apk").exec();
                RootUtil.setPermissions(644, Resources.DATA_DIR + "/IconifyComponent" + mOverlayName + ".apk");
            }
        }

        // Clean temp directory
        Shell.cmd("rm -rf " + Resources.TEMP_DIR).exec();
        Shell.cmd("rm -rf " + Resources.DATA_DIR + "/Overlays").exec();
    }

    private static void moveOverlaysToCache() {
        Shell.cmd("mv -f \"" + Resources.DATA_DIR + "/Overlays/" + mPackage + "/" + mOverlayName + "\" \"" + Resources.TEMP_CACHE_DIR + "/" + mPackage + "/" + mOverlayName + "\"").exec().isSuccess();
    }

    private static boolean createManifestResource(String overlayName, String source) {
        Shell.cmd("mkdir -p " + source + "/res").exec();
        String[] values = {"values", "values-land", "values-night"};

        for (int i = 0; i < 3; i++) {
            Shell.cmd("mkdir -p " + source + "/res/" + values[i]).exec();
            Shell.cmd("printf '" + mResource[i] + "' > " + source + "/res/" + values[i] + "/iconify.xml;").exec();
        }

        String category = OverlayUtil.getCategory(overlayName);
        List<String> module = new ArrayList<>();
        module.add("printf '<?xml version=\"1.0\" encoding=\"utf-8\" ?>");
        module.add("<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\" android:versionName=\"v1.0\" package=\"IconifyComponent" + overlayName + ".overlay\">");
        module.add("\\t<uses-sdk android:minSdkVersion=\"" + BuildConfig.MIN_SDK_VERSION + "\" android:targetSdkVersion=\"" + Build.VERSION.SDK_INT + "\" />");
        module.add("\\t<overlay android:category=\"" + category + "\" android:priority=\"1\" android:targetPackage=\"" + mPackage + "\" />");
        module.add("\\t<application android:allowBackup=\"false\" android:hasCode=\"false\" />");
        module.add("</manifest>' > " + source + "/AndroidManifest.xml;");

        Shell.Result result = Shell.cmd(String.join("\\n", module)).exec();

        if (result.isSuccess())
            Log.i(TAG + " - Manifest", "Successfully created manifest for " + overlayName);
        else {
            Log.e(TAG + " - Manifest", "Failed to create manifest for " + overlayName + '\n' + String.join("\n", result.getOut()));
            writeLog(TAG + " - Manifest", "Failed to create manifest for " + overlayName, result.getOut());
        }

        return !result.isSuccess();
    }
}

File: app/src/main/java/com/drdisagree/iconify/utils/overlay/compiler/DynamicCompiler.kt
package com.drdisagree.iconify.utils.overlay.compiler

import android.util.Log
import com.drdisagree.iconify.data.common.Const.FRAMEWORK_PACKAGE
import com.drdisagree.iconify.data.common.Const.LAUNCHER3_PACKAGE
import com.drdisagree.iconify.data.common.Const.PIXEL_LAUNCHER_PACKAGE
import com.drdisagree.iconify.data.common.Const.SETTINGS_PACKAGE
import com.drdisagree.iconify.data.common.Const.SYSTEMUI_PACKAGE
import com.drdisagree.iconify.data.common.Resources.BACKUP_DIR
import com.drdisagree.iconify.data.common.Resources.DATA_DIR
import com.drdisagree.iconify.data.common.Resources.OVERLAY_DIR
import com.drdisagree.iconify.data.common.Resources.SIGNED_DIR
import com.drdisagree.iconify.data.common.Resources.SYSTEM_OVERLAY_DIR
import com.drdisagree.iconify.data.common.Resources.TEMP_CACHE_DIR
import com.drdisagree.iconify.data.common.Resources.TEMP_DIR
import com.drdisagree.iconify.data.common.Resources.TEMP_OVERLAY_DIR
import com.drdisagree.iconify.data.common.Resources.UNSIGNED_DIR
import com.drdisagree.iconify.data.common.Resources.UNSIGNED_UNALIGNED_DIR
import com.drdisagree.iconify.utils.FileUtils.copyAssets
import com.drdisagree.iconify.utils.RootUtils.setPermissions
import com.drdisagree.iconify.utils.SystemUtils.mountRO
import com.drdisagree.iconify.utils.SystemUtils.mountRW
import com.drdisagree.iconify.utils.helper.BinaryInstaller.symLinkBinaries
import com.drdisagree.iconify.utils.overlay.OverlayUtils.disableOverlays
import com.drdisagree.iconify.utils.overlay.OverlayUtils.enableOverlays
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager.ResourceType
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager.generateXmlStructureForAllResources
import com.topjohnwu.superuser.Shell
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.IOException

object DynamicCompiler {

    private val TAG = DynamicCompiler::class.java.simpleName
    private var mForce = false
    private var mPackage: String? = null
    private var mOverlayName: String? = null
    private val mResource: MutableMap<ResourceType, ArrayList<String>> = mutableMapOf()

    @JvmOverloads
    @Throws(IOException::class)
    suspend fun buildDynamicOverlay(
        force: Boolean = true,
        overlaysToUpdate: List<String>? = null
    ): Boolean {
        mForce = force

        try {
            Shell.cmd("mkdir -p $BACKUP_DIR").exec()

            val resourcesMap = generateXmlStructureForAllResources(overlaysToUpdate)

            // Create overlay for each package
            for (packageName in resourcesMap.keys) {
                Log.i(TAG, packageName)
                mPackage = packageName

                mResource.clear()
                mResource[ResourceType.PORTRAIT] =
                    ArrayList(resourcesMap[packageName]!![ResourceType.PORTRAIT]!!)
                resourcesMap[packageName]!![ResourceType.LANDSCAPE]?.let {
                    mResource[ResourceType.LANDSCAPE] = ArrayList(it)
                }
                resourcesMap[packageName]!![ResourceType.NIGHT]?.let {
                    mResource[ResourceType.NIGHT] = ArrayList(it)
                }

                mOverlayName = getOverlayName(packageName)

                preExecute()
                moveOverlaysToCache()

                // Create AndroidManifest.xml
                if (createManifestResource(
                        mOverlayName,
                        mPackage,
                        "$TEMP_CACHE_DIR/$mPackage/$mOverlayName"
                    )
                ) {
                    Log.e(TAG, "Failed to create Manifest for $mOverlayName! Exiting...")
                    postExecute(true)
                    return true
                }

                // Build APK using AAPT
                if (OverlayCompiler.runAapt(
                        "$TEMP_CACHE_DIR/$mPackage/$mOverlayName",
                        mPackage
                    )
                ) {
                    Log.e(TAG, "Failed to build $mOverlayName! Exiting...")
                    postExecute(true)
                    return true
                }

                // ZipAlign the APK
                if (OverlayCompiler.zipAlign("$UNSIGNED_UNALIGNED_DIR/$mOverlayName-unsigned-unaligned.apk")) {
                    Log.e(
                        TAG,
                        "Failed to align $mOverlayName-unsigned-unaligned.apk! Exiting..."
                    )
                    postExecute(true)
                    return true
                }

                // Sign the APK
                if (OverlayCompiler.apkSigner("$UNSIGNED_DIR/$mOverlayName-unsigned.apk")) {
                    Log.e(TAG, "Failed to sign $mOverlayName-unsigned.apk! Exiting...")
                    postExecute(true)
                    return true
                }

                postExecute(false)
            }

            if (mForce) {
                Shell.cmd("rm -rf $BACKUP_DIR").exec()
                val overlayPackageNames = resourcesMap.keys
                    .map { "IconifyComponent${getOverlayName(it)}.overlay" }
                    .toTypedArray()

                // Disable the overlays in case they are already enabled
                disableOverlays(*overlayPackageNames)

                // Install from files dir
                for (packageName in resourcesMap.keys) {
                    val apkName = "IconifyComponent${getOverlayName(packageName)}.apk"

                    Shell.cmd(
                        "pm install -r $DATA_DIR/$apkName",
                        "rm -rf $DATA_DIR/$apkName"
                    ).exec()
                }

                // Move to system overlay dir
                mountRW()
                for (packageName in resourcesMap.keys) {
                    val apkName = "IconifyComponent${getOverlayName(packageName)}.apk"

                    Shell.cmd("cp -rf $SIGNED_DIR/$apkName $SYSTEM_OVERLAY_DIR/$apkName").exec()
                    setPermissions(644, "/system/product/overlay/$apkName")
                }
                mountRO()

                // Enable the overlays
                enableOverlays(*overlayPackageNames)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to build overlay! Exiting...", e)
            postExecute(true)
            return true
        }

        return false
    }

    @Throws(IOException::class)
    private suspend fun preExecute() = withContext(Dispatchers.IO) {
        // Create symbolic link
        symLinkBinaries()

        // Clean data directory
        Shell.cmd(
            "rm -rf $TEMP_DIR",
            "rm -rf $DATA_DIR/Overlays"
        ).exec()

        // Extract overlay from assets
        copyAssets("Overlays/$mPackage/$mOverlayName")

        // Create temp directory
        Shell.cmd(
            "rm -rf $TEMP_DIR; mkdir -p $TEMP_DIR",
            "mkdir -p $TEMP_OVERLAY_DIR",
            "mkdir -p $TEMP_CACHE_DIR",
            "mkdir -p $UNSIGNED_UNALIGNED_DIR",
            "mkdir -p $UNSIGNED_DIR",
            "mkdir -p $SIGNED_DIR",
            "mkdir -p $TEMP_CACHE_DIR/$mPackage/",
            "mkdir -p $BACKUP_DIR"
        ).exec()
    }

    private fun postExecute(hasErroredOut: Boolean) {
        if (!hasErroredOut) {
            // Move all generated overlays to module
            Shell.cmd(
                "cp -rf $SIGNED_DIR/IconifyComponent$mOverlayName.apk $OVERLAY_DIR/IconifyComponent$mOverlayName.apk"
            ).exec()
            setPermissions(644, "$OVERLAY_DIR/IconifyComponent$mOverlayName.apk")
            Shell.cmd(
                "cp -rf $SIGNED_DIR/IconifyComponent$mOverlayName.apk $BACKUP_DIR/IconifyComponent$mOverlayName.apk"
            ).exec()

            // Move to files dir
            if (mForce) {
                Shell.cmd(
                    "cp -rf $SIGNED_DIR/IconifyComponent$mOverlayName.apk $DATA_DIR/IconifyComponent$mOverlayName.apk"
                ).exec()
                setPermissions(644, "$DATA_DIR/IconifyComponent$mOverlayName.apk")
            }
        }

        // Clean temp directory
        Shell.cmd(
            "rm -rf $TEMP_DIR",
            "rm -rf $DATA_DIR/Overlays"
        ).exec()
    }

    private fun moveOverlaysToCache() {
        Shell.cmd(
            "mv -f \"$DATA_DIR/Overlays/$mPackage/$mOverlayName\" \"$TEMP_CACHE_DIR/$mPackage/$mOverlayName\""
        ).exec().isSuccess
    }

    private fun createManifestResource(
        overlayName: String?,
        targetPackage: String?,
        source: String
    ): Boolean {
        Shell.cmd("mkdir -p $source/res").exec()

        val resourceTypes = arrayOf(
            ResourceType.PORTRAIT to "values",
            ResourceType.LANDSCAPE to "values-land",
            ResourceType.NIGHT to "values-night"
        )

        resourceTypes.forEach { (resourceType, values) ->
            val dirPath = "$source/res/${values}"
            val filePath = "$source/res/${values}/iconify.xml"
            val resourceList = mResource[resourceType]?.let { ArrayList(it) }

            if (!resourceList.isNullOrEmpty()) {
                val commands = mutableListOf(
                    "mkdir -p $dirPath",            // Create the directory
                    "rm -f $filePath",              // Remove the file if it exists
                    "touch $filePath"               // Create an empty file
                ).apply {
                    add("echo '${resourceList.joinToString("\n")}' >> $filePath")
                }

                Shell.cmd(*commands.toTypedArray()).exec()
            }
        }

        return OverlayCompiler.createManifest(overlayName, targetPackage, source)
    }

    private fun getOverlayName(packageName: String): String {
        return when (packageName) {
            FRAMEWORK_PACKAGE -> "Dynamic1"
            SYSTEMUI_PACKAGE -> "Dynamic2"
            PIXEL_LAUNCHER_PACKAGE -> "Dynamic3"
            LAUNCHER3_PACKAGE -> "Dynamic4"
            SETTINGS_PACKAGE -> "Dynamic5"
            else -> throw Exception("Unknown package: $packageName")
        }
    }
}
